File: libsrc/units/NondimElastic.cc
// -*- C++ -*-
//
// ======================================================================
//
//                           Brad T. Aagaard
//                        U.S. Geological Survey
//
// {LicenseText}
//
// ======================================================================
//

#include <portinfo>

#include "NondimElastic.hh" // implementation of class methods

#include <sstream> // USES std::ostringstream
#include <stdexcept> // USES std::runtime_error
#include <assert.h> // USES assert()

// ----------------------------------------------------------------------
// Default constructor
spatialdata::units::NondimElastic::NondimElastic(void) :
  _length(1.0),
  _pressure(1.0),
  _time(1.0),
  _density(1.0)
{ // constructor
} // constructor

// ----------------------------------------------------------------------
// Default destructor
spatialdata::units::NondimElastic::~NondimElastic(void)
{ // destructor
} // destructor

// ----------------------------------------------------------------------
// Copy constructor.
spatialdata::units::NondimElastic::NondimElastic(const NondimElastic& dim) :
  _length(dim._length),
  _pressure(dim._pressure),
  _time(dim._time),
  _density(dim._density)
{ // copy constructor
} // copy constructor

// ----------------------------------------------------------------------
// Assignment operator.
const spatialdata::units::NondimElastic& 
spatialdata::units::NondimElastic::operator=(const NondimElastic& dim)
{ // operator=
  if (this != &dim) {
    _length = dim._length;
    _pressure = dim._pressure;
    _time = dim._time;
    _density = dim._density;
  } // if

  return *this;
} // operator=

// ----------------------------------------------------------------------
// Set value to nondimensionalize length scale in meters (SI units).
void
spatialdata::units::NondimElastic::lengthScale(const double value)
{ // lengthScale
  if (value <= 0.0) {
    std::ostringstream msg;
    msg << "Length scale (" << value << ") must be positive.";
    throw std::runtime_error(msg.str());
  } // if
  _length = value;
} // lengthScale

// ----------------------------------------------------------------------
// Set value to nondimensionalize pressure scale in Pascals (SI units).
void
spatialdata::units::NondimElastic::pressureScale(const double value)
{ // pressureScale
  if (value <= 0.0) {
    std::ostringstream msg;
    msg << "Pressure scale (" << value << ") must be positive.";
    throw std::runtime_error(msg.str());
  } // if
  _pressure = value;
} // pressureScale

// ----------------------------------------------------------------------
// Set value to nondimensionalize time scale in seconds (SI units).
void
spatialdata::units::NondimElastic::timeScale(const double value)
{ // timeScale
  if (value <= 0.0) {
    std::ostringstream msg;
    msg << "Time scale (" << value << ") must be positive.";
    throw std::runtime_error(msg.str());
  } // if
  _time = value;
} // timeScale

// ----------------------------------------------------------------------
// Set value to nondimensionalize density scale in kg/m^3 (SI units).
void
spatialdata::units::NondimElastic::densityScale(const double value)
{ // densityScale
  if (value <= 0.0) {
    std::ostringstream msg;
    msg << "Density scale (" << value << ") must be positive.";
    throw std::runtime_error(msg.str());
  } // if
  _density = value;
} // densityScale

// ----------------------------------------------------------------------
// Make length scales dimensionless.
void
spatialdata::units::NondimElastic::utonLength(double* const values,
					      const int nvalues) const
{ // utonLength
  assert( (0 < nvalues && 0 != values) ||
	  (0 == nvalues && 0 == values) );

  for (int i=0; i < nvalues; ++i)
    values[i] /= _length;
} // utonLength

// ----------------------------------------------------------------------
// Add dimensions to length scale.
void
spatialdata::units::NondimElastic::ntouLength(double* const values,
					      const int nvalues) const
{ // ntouLength
  assert( (0 < nvalues && 0 != values) ||
	  (0 == nvalues && 0 == values) );

  for (int i=0; i < nvalues; ++i)
    values[i] *= _length;
} // ntouLength

// ----------------------------------------------------------------------
// Make pressure scales dimensionless.
void
spatialdata::units::NondimElastic::utonPressure(double* const values,
						const int nvalues) const
{ // utonPressure
  assert( (0 < nvalues && 0 != values) ||
	  (0 == nvalues && 0 == values) );

  for (int i=0; i < nvalues; ++i)
    values[i] /= _pressure;
} // utonPressure

// ----------------------------------------------------------------------
// Add dimensions to pressure scale.
void
spatialdata::units::NondimElastic::ntouPressure(double* const values,
						const int nvalues) const
{ // ntouPressure
  assert( (0 < nvalues && 0 != values) ||
	  (0 == nvalues && 0 == values) );

  for (int i=0; i < nvalues; ++i)
    values[i] *= _pressure;
} // ntouPressure

// ----------------------------------------------------------------------
// Make time scales dimensionless.
void
spatialdata::units::NondimElastic::utonTime(double* const values,
					    const int nvalues) const
{ // utonTime
  assert( (0 < nvalues && 0 != values) ||
	  (0 == nvalues && 0 == values) );

  for (int i=0; i < nvalues; ++i)
    values[i] /= _time;
} // utonTime

// ----------------------------------------------------------------------
// Add dimensions to time scale.
void
spatialdata::units::NondimElastic::ntouTime(double* const values,
					    const int nvalues) const
{ // ntouTime
  assert( (0 < nvalues && 0 != values) ||
	  (0 == nvalues && 0 == values) );

  for (int i=0; i < nvalues; ++i)
    values[i] *= _time;
} // ntouTime

// ----------------------------------------------------------------------
// Make density scales dimensionless.
void
spatialdata::units::NondimElastic::utonDensity(double* const values,
					       const int nvalues) const
{ // utonDensity
  assert( (0 < nvalues && 0 != values) ||
	  (0 == nvalues && 0 == values) );

  for (int i=0; i < nvalues; ++i)
    values[i] /= _density;
} // utonDensity

// ----------------------------------------------------------------------
// Add dimensions to density scale.
void
spatialdata::units::NondimElastic::ntouDensity(double* const values,
					       const int nvalues) const
{ // ntouDensity
  assert( (0 < nvalues && 0 != values) ||
	  (0 == nvalues && 0 == values) );

  for (int i=0; i < nvalues; ++i)
    values[i] *= _density;
} // ntouDensity


// End of file 

File: spatialdata/units/NondimElastic.py
#!/usr/bin/env python
#
# ----------------------------------------------------------------------
#
#                           Brad T. Aagaard
#                        U.S. Geological Survey
#
# <LicenseText>
#
# ----------------------------------------------------------------------
#

## @file spatialdata/units/NondimElastic.py
##
## @brief Python manager for nondimensionalizing elasticity problems.
##
## Factory: nondimensional

from pyre.components.Component import Component

# NondimElastic class
class NondimElastic(Component):
  """
  Python manager for nondimensionalizing elasticity problems.

  Factory: nondimensional
  """

  # INVENTORY //////////////////////////////////////////////////////////

  class Inventory(Component.Inventory):
    """Python object for managing NondimElastic facilities and properties."""

    ## @class Inventory
    ## Python object for managing NondimElastic facilities and properties.
    ##
    ## \b Properties
    ## @li None
    ##
    ## \b Facilities
    ## @li None

  # PUBLIC METHODS /////////////////////////////////////////////////////

  def __init__(self, name="nondimelastic"):
    """
    Constructor.
    """
    Component.__init__(self, name, facility="nondimelastic")

    self.cppHandle = None
    from pyre.units.length import meter
    from pyre.units.pressure import pascal
    from pyre.units.time import second
    from pyre.units.mass import kilogram

    self._length = 1.0*meter
    self._pressure = 1.0*pascal
    self._time = 1.0*second
    self._density = 1.0*kilogram/meter**3
    return


  def initialize(self):
    """
    Initialize coordinate system.
    """
    self._createCppHandle()
    self.cppHandle.length = self._length.value
    self.cppHandle.pressure = self._pressure.value
    self.cppHandle.time = self._time.value
    self.cppHandle.density = self._density.value
    return


  def utonLength(self, value):
    """
    Make length scale dimensionless.
    """
    return value / self._length


  def ntouLength(self, value):
    """
    Add dimensions to length scale.
    """
    return value * self._length


  def utonPressure(self, value):
    """
    Make pressure scale dimensionless.
    """
    return value / self._pressure


  def ntouPressure(self, value):
    """
    Add dimensions to pressure scale.
    """
    return value * self._pressure


  def utonTime(self, value):
    """
    Make time scale dimensionless.
    """
    return value / self._time


  def ntouTime(self, value):
    """
    Add dimensions to time scale.
    """
    return value * self._time


  def utonDensity(self, value):
    """
    Make density scale dimensionless.
    """
    return value / self._density


  def ntouDensity(self, value):
    """
    Add dimensions to density scale.
    """
    return value * self._density


  # PRIVATE METHODS ////////////////////////////////////////////////////

  def _configure(self):
    """
    Setup members using inventory.
    """
    Component._configure(self)
    return


  def _createCppHandle(self):
    """
    Create handle to C++ object.
    """
    if None == self.cppHandle:
      import spatialdata.units.units as bindings
      self.cppHandle = bindings.NondimElastic()
    return


# FACTORIES ////////////////////////////////////////////////////////////

def nondimensional():
  """
  Factory associated with NondimElastic.
  """
  return NondimElastic()


# End of file 
